### ISSUE TITLE
Enhancement proposal: code refactoring

### ISSUE THREAD

**Issue by lbteixeira** [SALib/SALib#320]
Hello, guys.

I was checking this project today, and I found it really interesting. I've been making some sensitivity analysis during my PhD studies, so this project is really useful for me.

Checking the code, I realized that perhaps I could contribute someway. I would like to propose you a refactor of the code to make it cleaner and more readable. Here are some things I could do:

1. Change some variables' names to increase readability
2. Make sure the code is using PEP-8 style (for example, 79 characters per line, naming conventions, etc)
3. Abstract functions from the existing ones. The idea is to increase the test coverage and to increase readability, writing small functions that do only one thing
4. Write extra unit tests
5. Eliminate hard coding, creating new default variables for the functions (similar to what I did in issue #316 today)
6. Add annotations to the variables showing the expected types

My idea is to start small and to do only one change first. Then I would submit it and wait for your feedback.

How do you feel about that?

**Comment by willu47** [comment]
Hi @lbteixeira. Many thanks for the suggestion. And thanks for the initiative and enthusiasm! I think these points would make a nice addition to the `CONTRIBUTING.md` guidelines.  We do not currently have a code-style section, and writing up a clear set of guidelines for contributors would be useful for the SALib community.

Code style can be contentious. For example, 79 characters is rather limiting from my point of view. Also, many of these aspects can be automated, and we are already using git hooks via the `pre-commit` to enforce certain aspects of code style.  And please, no automated-linting pull requests that reformat the entire code base...

Increasing test coverage is welcome as there are some parts of the package which are currently only covered by regression tests. So digging into the individual methods and targeted refactoring would be good. Ideally the tests should be useful to those reading them, explaining what the code should do, so examples from text books translated to the code are really handy.

**Comment by lbteixeira** [comment]
Hello, @willu47. Thanks for your reply.

I agree 79 characters isn't always nice, and this isn't a real concern I have while writing code, although some developers get crazy if you don't follow this aspect of PEP-8. I'm happy to know you are more flexible about this. Anyway, the parts of the code I checked so far are almost entirely conforming to this 79 characters standard, so this isn't a main issue. And for sure, no automated-linting...

I'll suggest you that I start by working at the subpackage sample/morris, as this is more connected to what I did so far during my PhD. I'll make as many commits as I can, so you can track in details what I did and see what you like or not. Once we agree on the standards to follow, I can update the `CONTRIBUTING.md` file.

Another related topic: I saw that some projects have a permanent branch called, say, Develop. The idea is that everyone contributing will have a local copy of Master and Develop. So, the usual way to contribute would be:

1. Everyday, before starting to work, update the local repos (Master and Develop)
2. Before modifying a feature, create a new branch from the local Develop repo
3. Once the feature is done, switch back to Develop and pull the latest changes.
4. Merge the branch from step 2 into the local Develop
5. Push the changes to the remote Develop repository
6. Check for errors and eventually merge everything on the Develop repo
7. Once a new version of Develop is worth a new release, push it to the remote Master.

The idea is to create an extra layer between the developments going on and the Master. This flow is more or less represented in the image below:
![image](https://user-images.githubusercontent.com/10145845/84940011-81cb4780-b0df-11ea-9b13-118feea98004.jpg)

What are your thoughts about this? Could be useful or is it unnecessarily complex?
I would also like to hear from you about my suggestion regarding the subpackage sample/morris.

**Comment by ConnectedSystems** [comment]
Hi @lbteixeira 

You might also be interested in following the discussion in https://github.com/SALib/SALib/issues/234

**Comment by lbteixeira** [comment]
Hi, @ConnectedSystems 

Very interesting discussion going on in #234. I'll take some time soon to read the code in more details, but I would be glad to help if I can.

Once I finish this first round of refactoring with Morris module, I'll take a look at #265 too

**Comment by lbteixeira** [comment]
Hi, @ConnectedSystems 

One thing I noticed about the `sample` function in the `__init__.py` module of the subpackage `morris` is that it has an `if` statement that directs the calculation of the samples depending on what comes as an input with the `problem` dictionary, what is perfectly fine.

Nevertheless, I don't believe we need both functions `_sample_oat` and `_sample_groups`, since the latter is just a generalization of the former. If we make the number of groups equal to the number of input variables by default, the method `_sample_groups` will make the calculations as there were no groups at all. On the other hand, if the user explicitly define the groups, it'll also work.

Therefore, I'm going to delete the function `_sample_oat` and make the necessary changes so as everything continues to work fine, ok?

**Comment by lbteixeira** [comment]
I'm closing this issue since there is already a PR related to it. @willu47, as soon as you merge #322 I'll open a new issue to contribute more code.
